### PR TITLE
IGEN DTSD422-D3 Smartmeter support

### PR DESCRIPTION
**Pull request by ixs**
Add support for the DTSD422-D3 smartmeter which is normaly used to get grid/consumption information for the solarman platform.

### CODE REVIEW

File: deye_sensors_igen_dtsd422.py [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+class SingleDirectionRegisterSensor(SingleRegisterSensor):
+    """
+    Power meter sensor with value stored as 10-bit integer in a single Modbus register and the first byte
+    indicating power direction.
+    """
+
+    def read_value(self, registers: dict[int, int]):
+        if self.reg_address in registers:
+            reg_value = (
+                int.from_bytes(registers[self.reg_address], "big", signed=self.signed)
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:48]: This conversion extracts the magnitude, which is always positive, hence, `signed` should be set to `False`

File: deye_sensors_igen_dtsd422.py [hunk 2]
```diff
@@ -0,0 +1,703 @@
...
+            # If highest bit is set, we've got a negative value
+            if bool(
+                (
+                    int.from_bytes(
+                        registers[self.reg_address], "big", signed=self.signed
+                    )
+                    & 0x8000
+                )
+                >> 15
+            ):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:60]: This conversion can be simplified.
```
bool(registers[self.reg_address][0] & 0x80)
```

File: deye_sensors_igen_dtsd422.py [hunk 3]
```diff
@@ -0,0 +1,703 @@
...
+    """
+
+    def read_value(self, registers: dict[int, int]):
+        high_word_reg_address = self.reg_address
+        low_word_reg_address = self.reg_address + 1
+        if low_word_reg_address in registers and high_word_reg_address in registers:
+            low_word = registers[low_word_reg_address]
+            high_word = registers[high_word_reg_address]
+            reg_value = (
+                int.from_bytes(high_word + low_word, "big", signed=self.signed)
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:82]: This conversion extracts the magnitude, which is always positive, hence, `signed` should be set to `False`

File: deye_sensors_igen_dtsd422.py [hunk 4]
```diff
@@ -0,0 +1,703 @@
...
+                & 0x7FFFFFFF
+            )
+            # If highest bit is set, we've got a negative value
+            if bool(
+                (
+                    int.from_bytes(high_word + low_word, "big", signed=self.signed)
+                    & 0x80000000
+                )
+                >> 31
+            ):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:92]: This conversion can be simplified.
```
bool(registers[self.reg_address][0] & 0x80)
```

File: deye_sensors_igen_dtsd422.py [hunk 5]
```diff
@@ -0,0 +1,703 @@
...
+#
+# Initial IGEN DTSD-422-D3 support.
+#
+# Todo:
+# * Daily Values for Positive and Negative energy are missing, can't find them.
+# * Is the direction calculation handled correctly? Or is something wrong here?
+# * Need to verify data is still read correctly when double reg values exceed one
+#   register
+# * There are some more informational registers, e.g. Manufacturing Date, serial etc. which
+#   we're ignoring for now. Maybe add?
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:35]: What about this? Would you like to tackle any of those todos before merging this PR?

**Review by kbialek — CHANGES_REQUESTED** [review]
Awesome contribution. Great job!

File: deye_sensors_igen_dtsd422.py (hunk 1 above)
Comment (ixs) [reply deye_sensors_igen_dtsd422.py:48]: Thanks. First time I'm bit-juggling with modbus...

File: deye_sensors_igen_dtsd422.py [hunk 6]
```diff
@@ -0,0 +1,687 @@
...
+#
+
+
+class SingleDirectionRegisterSensor(SingleRegisterSensor):
+    """
+    Power meter sensor with value stored as 10-bit integer in a single Modbus register and the first byte
+    indicating power direction.
+    """
+
+    def read_value(self, registers: dict[int, int]):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:44]: Please change `registers` type hint to `dict[int, bytearray]`
I've realized, that it's wrong when reviewing your PR.
The fix is already in the `main` branch, so you should rebase your PR.

File: deye_sensors_igen_dtsd422.py [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+
+
+class DoubleDirectionRegisterSensor(DoubleRegisterSensor):
+    """
+    Power meter sensor with value stored as 32-bit integer in two Modbus registers and the energy flow direction
+    stored in the first Modbus register. High and Low words are swapped compared to the normal DoubleRegisterSensor.
+    Highest bit gives energy direction.
+    """
+
+    def read_value(self, registers: dict[int, int]):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:66]: As above, please update the type hint

File: deye_sensors_igen_dtsd422.py [hunk 8]
```diff
@@ -0,0 +1,687 @@
...
+# Todo:
+# * Daily Values for Positive and Negative energy are missing, can't find them.
+# * Need to verify data is still read correctly when double reg values exceed one
+#   register
+# * There are some more informational registers, e.g. Manufacturing Date, serial etc. which
+#   we're ignoring for now. Maybe add?
+#
+
+
+class SingleDirectionRegisterSensor(SingleRegisterSensor):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:38]: This class can be renamed to `SignedMagnitudeSingleRegisterSensor` and moved into `deye_sensor.py`
Please add corresponding conversion tests in `deye_sensor_test.py`

File: deye_sensors_igen_dtsd422.py [hunk 9]
```diff
@@ -0,0 +1,687 @@
...
+            # If highest bit is set, we've got a negative value
+            if bool(registers[self.reg_address][0] & 0x80):
+                return -1 * reg_value * self.factor + self.offset
+            else:
+                return reg_value * self.factor + self.offset
+        else:
+            return None
+
+
+class DoubleDirectionRegisterSensor(DoubleRegisterSensor):
```
Comment (kbialek) [inline deye_sensors_igen_dtsd422.py:59]: This class can be renamed to `SignedMagnitudeDoubleRegisterSensor` and moved into `deye_sensor.py`
Please add corresponding conversion tests in `deye_sensor_test.py`

**Review by kbialek — CHANGES_REQUESTED** [review]
I've added few more remarks. Mainly focused on better code organization and tests.

File: deye_sensors_igen_dtsd422.py (hunk 5 above)
Comment (kbialek) [reply deye_sensors_igen_dtsd422.py]: Exactly, they do not bring much value, maybe besides of printing them in the log at startup.

File: deye_sensor.py [hunk 10]
```diff
@@ -114,6 +114,52 @@ def get_registers(self) -> list[int]:
         return [self.reg_address, self.reg_address+1]
 
 
+class SignedMagnitudeSingleRegisterSensor(SingleRegisterSensor):
+    """
+    Power meter sensor with value stored as 10-bit integer in a single Modbus register and the first byte
+    indicating power direction.
```
Comment (kbialek) [inline deye_sensor.py:120]: Alternative comment proposal
```
Reads single Modbus register as signed, fixed point 15-bits long value. The most significant bit encodes the sign. 
```

File: deye_sensor.py [hunk 11]
```diff
@@ -114,6 +114,52 @@ def get_registers(self) -> list[int]:
...
+                return reg_value * self.factor + self.offset
+        else:
+            return None
+
+
+class SignedMagnitudeDoubleRegisterSensor(DoubleRegisterSensor):
+    """
+    Power meter sensor with value stored as 32-bit integer in two Modbus registers and the energy flow direction
+    stored in the first Modbus register. High and Low words are swapped compared to the normal DoubleRegisterSensor.
+    Highest bit gives energy direction.
```
Comment (kbialek) [inline deye_sensor.py:142]: Alternative comment proposal
```
Reads double Modbus registers as signed, fixed point 31-bits long value. The most significant bit encodes the sign. 
```

**Review by kbialek — CHANGES_REQUESTED** [review]
The very last remark about the comments

**Review by kbialek — APPROVED** [review]

**Pull request merged**